src/mesh_analysis.rs

```diff
@@ -64,25 +64,59 @@ pub fn non_manifold_edges<'a>(
 /// Finds border vertex indices in a mesh edge collection
 /// A vertex is border when its edge's valency is 1
 #[allow(dead_code)]
-pub fn border_vertex_indices(edge_valencies: &EdgeCountMap) -> HashSet<u32> {
+pub fn border_vertex_indices(edge_sharing: &EdgeSharingMap) -> HashSet<u32> {
     let mut border_vertices = HashSet::new();
 
-    border_edges(edge_valencies).for_each(|edge| {
+    border_edges(edge_sharing).for_each(|edge| {
         border_vertices.insert(edge.vertices.0);
         border_vertices.insert(edge.vertices.1);
     });
     border_vertices
 }
 
+/// Finds continuous loops of border edges,
+/// starting from a random edge.
+/// The mesh may contain holes or islands,
+/// therefore it may have an unknown number of edge loops.
+/// If two edge loops meet at a single vertex, the result
+/// may be unpredictable and erratic.
+#[allow(dead_code)]
+pub fn border_edge_loops(edge_sharing: &EdgeSharingMap) -> Vec<Vec<UnorientedEdge>> {
+    let mut border_edges: Vec<_> = border_edges(edge_sharing).map(UnorientedEdge).collect();
+
+    let mut edge_loops: Vec<Vec<UnorientedEdge>> = Vec::new();
+
+    while !border_edges.is_empty() {
+        if let Some(edge) = border_edges.pop() {
+            let mut current_loop: Vec<UnorientedEdge> = vec![edge];
+
+            while current_loop.len() < 3
+                || !current_loop[current_loop.len() - 1].shares_vertex(current_loop[0])
+            {
+                for (i, other_edge) in border_edges.iter().enumerate() {
+                    if current_loop[current_loop.len() - 1].shares_vertex(*other_edge) {
+                        current_loop.push(border_edges.remove(i));
```

Comment:
Is this the source of erratic behavior from the doc comment? Overlapping loops can "steal" border edges from each other?

Reply:
no, that would be an error. see the picture - how would the edge loop look like in this case? and what should it look like?
![image](https://user-images.githubusercontent.com/3333537/67385050-97a95c80-f592-11e9-911d-acd88236d9c4.png)

Reply:
Well that's what I was trying to say. Currently it is implementation dependent: whatever border edge is encountered first changes the result for the rest of them.

Reply:
In this picture I see at least 8 unique edge loops... Not sure if you want to return all or not.